Added video sections, subsections, and videos page.  Stubbed video page.

dsjen opened the pull request:
I'm creating this PR to gather early feedback.  Please give it a once over if you have a moment.  I'm still working on adding documentation and writing both acceptance and unit tests, so I'm not anticipating much changing structurally.  Also, I'm still waiting on the final labels to add to the UI.

You'll find that I've pulled a lot of the course structure code into a mixin used by both the performance and engagement presenter and views.

Please let me know if you have any questions!

@brianhw @mulby @jab5569

brianhw commented on acceptance_tests/test_course_detail.py:
```diff
@@ -64,21 +64,29 @@ def _test_table(self):
+        }
+        if ENABLE_COURSE_API:
+            engagement['items'].append({
+                'title': 'How many students are interacting with my course?',
```
naive question: is this supposed to have the same title as the previous item above?

dsjen replied:
Good question!  This is a placeholder for the actual label.  I just changed it to "TBD" with a "TODO" so we don't forget about it.  Thank you!

brianhw commented on acceptance_tests/test_course_engagement.py:
```diff
@@ -108,3 +111,12 @@ def _test_engagement_table(self):
+class CourseEngagementVideoTests(CoursePerformancePageTestsMixin, WebAppTest):
+
+    def setUp(self):
+        super(CourseEngagementContentTests, self).setUp()
```
This needs to have the right class name:  s/Content/Video/

dsjen replied:
Thanks!

brianhw commented on analytics_dashboard/courses/exceptions.py:
```diff
@@ -26,16 +26,29 @@ class PermissionsRetrievalFailedError(PermissionsError):
-        super(NoAnswerSubmissionsError, self).__init__(*args, **kwargs)
-        self.message = 'No answers have been submitted for course {}.'.format(self.course_id)
+        super(BaseCourseError, self).__init__(*args, **kwargs)
+        self.message = self.message_template.format(self.course_id)
```
Where does message_template get set?

brianhw replied:
Check for None?  Or make a property that raises a NotImplementedException?

dsjen replied:
I like that!  I'll turn `message_template` into an `abstractproperty`.

brianhw replied:
Done.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+        super(CourseAPIPresenterMixin, self).__init__(course_id, timeout)
+        self.course_api_client = CourseStructureApiClient(settings.COURSE_API_URL, access_token)
+
+    def _structure(self):
```
nit: just a naming question.  Should this be a property, or should it instead be named _get_structure()?

dsjen replied:
`_get_structure()` seems clearer to me.  I'll update the name.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+        for index, collection in enumerate(collections):
+            self.build_module(index, collection, url_func)
+
+
```
nit: extra line

dsjen replied:
Eagle eyes!  :)

brianhw commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,3 +92,92 @@ def get_summary_and_trend_data(self):
+
+    def build_subsection_url_func(self, section_id):
+        """
+        Returns a function for creating the ungraded subsection URL.
```
nit: remove "ungraded"

dsjen replied:
Update--thanks!

brianhw commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,3 +92,92 @@ def get_summary_and_trend_data(self):
+
+    def post_process_adding_data_to_blocks(self, data, parent_block, child_block, url_func=None):
+        if url_func and data['start_views'] > 0:
+            data['url'] = url_func(parent_block, child_block)
```
Only videos with start_views will get a url?  Is it possible for a video to have start_views=0 (because people watched starting at an offset) but still want this?

dsjen replied:
This is a good question.  I'll ask Product about what the proper criteria will be.

brianhw commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -433,55 +439,64 @@ class CourseHome(CourseTemplateWithNavView):
+        engagement_items = {
+            'name': _('Engagement'),
+            'icon': 'fa-bar-chart',
+            'heading': _('What are students doing in my course?'),
```
(If you have to ask this question, you have to wonder what students *are* doing in your course!)

dsjen replied:
:)

brianhw commented:
The overall refactoring looks great, best I can tell.  Nice!

brianhw commented on analytics_dashboard/courses/views/__init__.py:
```diff
@@ -550,3 +565,68 @@ def _create_course_list(self, course_ids):
+            raise
+
+
+class CourseStructureMixin(object):
```
If/when you get around to writing docstrings for these mixins, it would be good to be clear about the order in which this mixin and the CourseStructureExceptionMixin class should be included in a class definition, if in fact it matters.  (I'm assuming that it does matter in terms of how the dispatch() function works.

brianhw replied:
Can this also get a docstring?

dsjen replied:
Of course!

brianhw commented on analytics_dashboard/courses/views/engagement.py:
```diff
@@ -53,3 +58,88 @@ def get_context_data(self, **kwargs):
+    template_name = 'courses/engagement_video_course.html'
+    page_name = 'engagement_videos'
+
+
```
nit: extra newline.

dsjen replied:
Thanks!  Updated.

dsjen commented:
@brianhw, please feel free to review this PR.  I've added unit and acceptance tests as well as updated documentation.  The acceptance tests are failing on Travis and I'm looking what's gone wrong; they are passing locally.

brianhw commented on acceptance_tests/mixins.py:
```diff
@@ -62,6 +62,9 @@ def assertTableColumnHeadingsEqual(self, table_selector, headings):
 
+    def assertTableColumns(self, selector, expected_column_headings):
+        self.assertTableColumnHeadingsEqual(selector, expected_column_headings)
+
```
I'm not clear what value this method provides.  Seems like the code should just call assertTableColumnHeadingsEqual() directly, no?

dsjen replied:
Good call!  This was part of a refactoring...

brianhw commented on acceptance_tests/test_course_engagement.py:
```diff
@@ -108,3 +119,87 @@ def _test_engagement_table(self):
+                       u'why.'
+    expected_table_columns = [u'Order', u'Section Name', u'Videos', u'Started Watching', u'Finished Watching',
+                              u'Completion Percentage']
+
```
Are some of these expected values actually defined in the original code?  If so, you could just import from there.

dsjen replied:
These are set in javascript, so unfortunately difficult to pull over.

brianhw replied:
Okay, thanks.

brianhw commented on acceptance_tests/test_course_engagement.py:
```diff
@@ -108,3 +119,87 @@ def _test_engagement_table(self):
+    # pylint: disable=line-too-long
+    expected_tooltip = u'Each bar shows the counts of complete and incomplete plays of the videos in that ' \
+                       u'subsection. Click on bars with low totals or a high incomplete rate to drill down and ' \
+                       u'understand why.'
```
curious why the pylint is needed -- this line looks short enough....

dsjen replied:
Thanks!  I'll remove them.  I originally had it as a very long line, but found it unreadable.

brianhw commented on acceptance_tests/test_course_performance.py:
```diff
@@ -111,11 +111,8 @@ def _build_submissions(self, blocks, problems):
-
     def assertBlockRows(self, blocks, include_children_count=True):
-        table = self.page.browser.find_element_by_css_selector('div[data-role="performance-table"]')
+        table = self.page.browser.find_element_by_css_selector('div[data-role="data-table"]')
```
nit: seems like the 'div[...]' could be pulled out into a constant, since it's used in a number of methods below.

dsjen replied:
Thanks, I'll pull that selector out.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+
+                if found_structure and subsection_id:
+                    found_structure = \
+                        [section for section in found_structure[0]['children'] if section['id'] == subsection_id]
```
Is it possible to have subsection_id and not have section_id?  This seems to assume that there is some kind of section_id that is involved in returning  found_structure[0].

brianhw replied:
Maybe there should be a check to see that if subsection_id != None, then section_id is also != None.  And then this last clause woudl have the subsection_id test indented within the "if section_id:" clause.

dsjen replied:
I think I can achieve that by nested the subsection part under the `if section_id` part, like:

```
            if found_structure:
                if section_id:
                    found_structure = [section for section in found_structure if section['id'] == section_id]

                    if found_structure and subsection_id:
                        found_structure = \
                            [section for section in found_structure[0]['children'] if section['id'] == subsection_id]
```

Is that what you were thinking?

brianhw replied:
That was the last sentence of what I was thinking.  There's still a check that could/should be done.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+
+    @abc.abstractmethod
+    def process_module_data(self, data):
+        pass
```
Is this the right place to document what this method should be used for?   It's not at all clear from the name what this is actually doing.  Is it modifying data, or modifying something else based on the data?

dsjen replied:
Thanks, I'll add documentation.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+    @abc.abstractmethod
+    def fetch_course_module_data(self):
+        """
+        Fetch course module data form the data API.  Use course_module_data for cached data.
```
Should one use course_module_data?  There's a _course_module_data below.  And if so, should this also have a leading underscore?  Perhaps not if it's being defined in a derived class.

dsjen replied:
Ah, I'll update the documentation to use the `_course_module_data()`.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+
+        return module_data
+
+    def module_id_to_data_id(self, data):
```
docstring?   What is the data being passed in?

dsjen replied:
Good point!  I updated the argument to `module` rather and added the following docstring: "Translates the course structure module to the ID used by the analytics data API."

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+        """ Retrieve a specific block (e.g. problem, video). """
+        block = self._get_structure()['blocks'][block_id]
+        block['name'] = block.pop('display_name')
+        return block
```
I assume that block is a copy, so that it's not blowing away 'name' with subsequent calls.

brianhw commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,3 +93,110 @@ def get_summary_and_trend_data(self):
+        # Change the id key name
+        if 'encoded_module_id' in video:
+            video_id = video.pop('encoded_module_id')
+            video['id'] = video_id
```
Doesn't look like video_id is ever used.

dsjen replied:
Consolidated to `video['id'] = video.pop('encoded_module_id')`.

brianhw commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,3 +93,110 @@ def get_summary_and_trend_data(self):
+        return build_url
+
+    def post_process_adding_data_to_blocks(self, data, parent_block, child_block, url_func=None):
+        data['url'] = url_func(parent_block, child_block)
```
Looks like this would not work if url_func=None.  So maybe make it non-optional?

dsjen replied:
Oops, I did in fact intend that to be optional.

brianhw commented on analytics_dashboard/courses/presenters/performance.py:
```diff
@@ -208,57 +188,33 @@ def assignment_types(self):
-            cache.set(key, problems)
-
+    def fetch_course_module_data(self):
+        # Get the problems from the API.  Use course_module_data() for cached data.
```
Maybe just note that this implements the abstract method, instead of the "Use..." sentence.  And maybe in the docstring.

dsjen replied:
Thanks!  Updated docstring.

brianhw commented on analytics_dashboard/courses/tests/factories.py:
```diff
@@ -160,3 +160,73 @@ def present_sections(self):
+        return video
+
+    @property
+    def present_sections(self):
```
nit:  I understand that "present" here is a verb, but because this is a property, I expect it to be a "thing".  So while most methods might be verb_object, this should probably just be noun:  "presented_sections".

dsjen replied:
Makes perfect sense.  I've updated to use "presented" now.

brianhw commented on analytics_dashboard/courses/tests/test_presenters.py:
```diff
@@ -40,9 +42,14 @@ def test_get_current_date(self):
 class CourseEngagementPresenterTests(SwitchMixin, TestCase):
+    SECTION_ID = 'i4x://edX/DemoX/chapter/9fca584977d04885bc911ea76a9ef29e'
+    SUBSECTION_ID = 'i4x://edX/DemoX/sequential/07bc32474380492cb34f76e5f9d9a135'
+    VIDEO_ID = 'i4x://edX/DemoX/video/0b9e39477cf34507a7a48f74be381fdd'
```
Out of curiosity, do you also have any tests for newer-opaque-keys-type IDs?  Not sure that the code here really depends on it anywhere, fortunately...

brianhw replied:
... except in the module_id_to_data_id level of things.

dsjen replied:
We don't have any tests at the video of problem level, but we do at the course ID level.

brianhw commented on analytics_dashboard/courses/tests/test_presenters.py:
```diff
@@ -115,6 +122,105 @@ def test_get_summary_and_trend_data_small(self, mock_activity):
+    def test_module_id_to_data_id(self):
+        expected_id = 'i4x-edX-DemoX-video-0b9e39477cf34507a7a48f74be381fdd'
+        module_id = 'i4x://edX/DemoX/video/0b9e39477cf34507a7a48f74be381fdd'
+        self.assertEqual(self.presenter.module_id_to_data_id({'id': module_id}), expected_id)
```
So this would be good to have an opaque-key- equivalent test here too.

dsjen replied:
Good call!  I added one in.

brianhw commented on analytics_dashboard/courses/tests/test_views/__init__.py:
```diff
@@ -263,3 +264,91 @@ def assertTertiaryNavs(self, nav, course_id):
+
+    def test_invalid_course(self):
+        """
+        The view should return HTTP 404 if the course is invalid.
```
Is this true?  Or does it depend on the implementation?

dsjen replied:
Ah, it *will*!  I'll update the doc.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+                child.update(data)
+
+    def post_process_adding_data_to_blocks(self, data, parent_block, child, url_func=None):
+        """ Override this if additional data is needed on the child block (e.g. problem part data). """
```
Can you say something about the args here, as to what is being passed in?

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,217 @@ def strip_time(s):
+
+    @abc.abstractmethod
+    def blocks_have_data(self, blocks):
+        pass
```
does this return true if any block has data, or if all have data?  answer should be in the docstring.

brianhw commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -1,15 +1,24 @@
+class CourseEngagementPresenter(CourseAPIPresenterMixin, BasePresenter):
     """
-    Presenter for the engagement page.
+    Presenter for the engagement content page.
```
By the way, this is probably not the right place (or right time) to be asking this, but these are very video-specific engagement "presenters", and yet the word "Video" doesn't actually show up in any of the class names.  There's a mention of "content" here, but isn't it possible (if not likely) that we would have other kinds of engagement with content that we would be displaying at some point in the near future?  Admittedly, there can be another refactor at that point, but it seems like you have done a lot here to generalize from video and problem.  Maybe there really isn't anything else that involves scanning the courseware in the same way, but I just wonder if the class names are accurate here.

dsjen replied:
You know, this is a very good point.  I could probably separate these into two engagement presenters--one for the activity and one for video.  This should be easy enough...I think.  I'll give it a shot.

brianhw commented on analytics_dashboard/courses/presenters/performance.py:
```diff
@@ -325,31 +245,30 @@ def assignments(self, assignment_type=None):
-            # removing the URL keeps navigation between the menu and bar chart consistent
-            if url_func and submission_collection['total_submissions'] > 0:
-                submission_collection['url'] = url_func(submission_collection)
+    def build_module(self, index, collection, url_func=None):
```
Just out of curiosity, what is the boundary between build_module and process_module_data?  This doesn't seem to use the latter.

dsjen replied:
I think `build_module` is poorly named.  It is used to build the aggregate data from the children modules into the collection/parent.  I've renamed it to `build_aggregate_parent_module` and added documentation.  I hope that helps!

brianhw commented on analytics_dashboard/courses/tests/test_views/test_engagement.py:
```diff
@@ -24,15 +39,41 @@ def assertPrimaryNav(self, nav, course_id):
         self.assertDictEqual(nav, expected)
 
+    def get_expected_secondary_nav(self, _course_id):
+        # override for each secondary pages
```
nit: s/pages/page

dsjen replied:
Thanks, updated!

brianhw commented on analytics_dashboard/courses/tests/test_views/test_engagement.py:
```diff
@@ -82,3 +123,110 @@ def assertValidMissingDataContext(self, context):
+
+    @httpretty.activate
+    @patch('courses.presenters.engagement.CourseEngagementPresenter.section', Mock(return_value=None))
+    def test_missing_subsections(self):
```
test_missing_section?

dsjen replied:
This is oddly correctly.  A section consists of many subsections.

brianhw replied:
But the mock is returning None for a call to section('Invalid'), so I don't see anything that's testing subsections.

dsjen replied:
Oops, you are absolutely right.  I've updated the test to check section! :)

brianhw commented:
Refactoring looks good -- only a few nits about naming.  And I'll make a future pass regarding docstrings.

brianhw commented:
FWIW, the Travis CI build has the same message I was encountering locally:  

Compressing... make: *** [test_python] Killed
The command "make validate" exited with 2.

Perhaps that contributed to other failures....

dsjen commented:
@brianhw Thank you for the great feedback!  I've updated the code based on your recommendations and added documentation for many of the methods in `CourseAPIPresenterMixin`.  Please let me know if there's anything else that you'd like addressed.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,254 @@ def strip_time(s):
+    @abc.abstractmethod
+    def fetch_course_module_data(self):
+        """
+        Fetch course module data form the data API.  Use _course_module_data() for cached data.
```
nit:  s/form/from/

dsjen replied:
Updated.

brianhw commented on analytics_dashboard/courses/presenters/engagement.py:
```diff
@@ -84,3 +93,113 @@ def get_summary_and_trend_data(self):
+        # calculates the percentages too
+        self.process_module_data(collection)
+
+        # removing the URL keeps navigation between the menu and bar chart consistent
```
I'm confused:  are you removing the URL here?

dsjen replied:
That comment is unclear.  I've updated it to "including the URL enables navigation to child pages" instead.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,254 @@ def strip_time(s):
+        """
+        Called by _course_module_data() to attach computed data (e.g. percentages, new IDs, etc.) to
+        data returned from the analytics data api.
+        """
```
nit: I like the name you suggest in the docstring much more than the actual name of the method:  "attach_computed_data(self, module_data)".  And the argument seems like it's really the "module_data", not just "data".

dsjen replied:
Thanks, I think that naming is clearer too--updated.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,254 @@ def strip_time(s):
+
+    @abc.abstractmethod
+    def blocks_have_data(self, blocks):
+        """ Returns whether blocks contains displayable data. """
```
nit: Still would like to know if this is "any" or "all"....

dsjen replied:
Ah, yes, this is "any" data.  I'll update the comment.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,254 @@ def strip_time(s):
+        return module['id']
+
+    def add_data_to_blocks(self, blocks, url_func=None):
+        """ Attaches data from the analytics data API to the course structure modules. """
```
nit: Comparing this with process_module_data, this seems to specifically add data about children.  So this might be more accurately named: "add_child_data_to_parent_blocks(self, parent_blocks, url_func=None):".

dsjen replied:
Agreed!  That name is definitely clearer.  Updated.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,254 @@ def strip_time(s):
+
+    @abc.abstractmethod
+    def build_aggregate_parent_module(self, index, collection, url_func=None):
+        """ Builds aggregate data from the child modules to the collection. """
```
nit: I didn't think of "builds" as something one can build from and build to.   But I'm starting to understand the code better, and now can understand better what this means.  But in first reading, I had no idea.  I think that if we had "attach_computed_data" below, this would be more accurately "attach_aggregated_data_to_parent".   You also introduce the variable "collection" in these, but the terms elsewhere are things like "module_data" or "parent".  I'd suggest the variable here is "parent".  And that the comment replaces "Builds" with "Adds" and replaces "collection" with "parent".

dsjen replied:
I really like that!  Updated.

brianhw commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -41,3 +48,254 @@ def strip_time(s):
+
+        return found_structure
+
+    def build_modules(self, collections, url_func=None):
```
If this is only used above, it could be prepended with an underscore.  I don't think it's part of the interface.

dsjen replied:
It turns out that this is used in the performance presenter too.

brianhw commented:
I reviewed the __init__ files again, and suggested some method renames that might help and some docstrings that would be good to add, if you think it's reasonable.  Otherwise :+1:

dsjen commented:
@brianhw thank you!  I've renamed the methods and added the docstrings.  That all made a lot of sense.
